### ISSUE TITLE
popstate event handlers not triggered in Firefox on hash change

### ISSUE THREAD

**Issue by rosenquist** [withastro/astro#10249]
### Astro Info

```block
Astro                    v4.4.5
Node                     v20.11.1
System                   macOS (arm64)
Package Manager          pnpm
Output                   static
Adapter                  none
```

### If this issue only occurs in one browser, which browser is a problem?

Firefox

### Describe the Bug

In Firefox, with astro view transitions enabled, clicking hash links to navigate to a specific part of a page doesn't trigger `popstate` event handlers. If I remove the `ViewTransitions` component, the event is triggered as expected.

`popstate` event handlers are triggered successfully on back/forward-button navigation.

I have added a stackblitz link for reproduction but something like this should demonstrate the issue:
```astro
---
import { ViewTransitions } from 'astro:transitions';
---
<html>
  <head>
    <ViewTransitions />
  </head>
  <body>
    <a href="#1234">link to title</a>
    <a href="#asdf">link to sub title</a>
    <h1 id="1234">Title</h1>
    <h2 id="asdf">Sub title</h2>
  </body>
</html>

<script>
window.onpopstate = (e) => {
  console.log('popstate event:', e);
};
</script>
```

### What's the expected result?

`popstate` event handlers should be triggered consistently in all major browsers.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-dbavaf?file=src%2Flayouts%2FLayout.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by martrapp** [comment]
Hi @rosenquist, thank you for addressing this issue.

An initial analysis shows that at least Chrome and Firefox differ in the way they emit popstate events, even when they run the same code. We had a similar issue with Safari before.  Getting all major browsers in line here seems to be a challenge.
I can confirm that the behavior of Firefox with view transitions is not consistent with either Firefox without view transitions nor Chrome.
Fixing this may require some major low-level effort that may not necessarily lead to complete success.
I will look into this, but also wanted to point out in advance that this might take a while ;-)

**Comment by Fryuni** [comment]
Just want to point it out that `popstate` is not required to fire in that case. Firefox is technically not incorrect here.

The specification defines the firing of `popstate` on:
- History navigation
- "Non-fragment synchronous navigation" (specifically restricted to "non-fragment" in the spec)
- Some cases of updating the document

A browser may also fire the `popstate` if it considers that the state has changed. This leads to an implementation detail in Chrome that triggers it, and I believe Safari as well.

The event you can and should rely on for changes in the fragment is the [`hashchange`](https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event)

**Comment by rosenquist** [comment]
@martrapp Thanks for looking into this issue.

@Fryuni Thanks for pointing this out. I originally used the `hashchange` event but it didn't trigger my event handler at all which is why I tried `popstate` instead, and using `popstate` actually did work in chrome and safari. Also, the `popstate` event handlers are triggered in firefox when you don't use `ViewTransitions`, so it seems like an issue related to the Astro view transitions implementation, not the browser.

Going back to trying the `hashchange` event instead, it turns out that it actually works if I comment out `ViewTransitions`. But it's currently not firing in any browser with `ViewTransitions` enabled.

So, in summary:
- `hashchange` event handlers are not being triggered in safari, chrome or firefox when `ViewTransitions` are enabled in my Astro project.
- `popstate` event handlers are not being triggered in firefox but are being triggered in safari and chrome, when `ViewTransitions` are enabled.
- Without `ViewTransitions`, both `popstate` and `hashchange` event handlers are triggered in safari, chrome and firefox.

**Comment by martrapp** [comment]
As @Fryuni pointed out, this is very thin ice. The Specs are cryptic at best and browsers implement them differently. I would love to craft a viable solution. Keep fingers crossed!

**Comment by martrapp** [comment]
Update from inside the rabbit hole:
As discussed [elsewhere](https://github.com/withastro/astro/issues/10309), we can't and do not want to mimic all navigation browser events with view transitions. There had been other issues with Firefox showing strange behavior when updating the current location.  I plan to add the missing popstate events right in the special cased code we already have here. Note that this does not aim on simulating events raised during non-view-transition navigation. But it should close the gap pointed out in this issue.
